mne/report/report.py

```diff
@@ -3166,6 +3175,30 @@ def _render_epochs(self, *, epochs, add_psd, add_projs, image_format,
 
         # PSD
         if add_psd:
+            if psd_signal_duration is None:  # Entire time range -> all epochs
+                epochs_for_psd = epochs  # Avoid creating a copy
+            else:  # Only a subset of epochs
+                epoch_duration = epochs.tmax - epochs.tmin
+                signal_duration = len(epochs) * epoch_duration
+                num_of_epochs_required = int(
+                    np.ceil(psd_signal_duration / epoch_duration)
+                )
+                if num_of_epochs_required > len(epochs):
+                    raise ValueError(
+                        f'You requested to calculate PSD on a duration of '
+                        f'{psd_signal_duration:.3f} sec, but all your epochs '
+                        f'are only {signal_duration:.3f} sec long'
+                    )
+                epochs_idx = np.arange(
+                    start=0,
+                    stop=len(epochs),
+                    step=int(np.floor(len(epochs) / num_of_epochs_required))
+                )
+                # We may have more epochs than necessary due to the floor();
+                # simply drop the last few ones in this case
```

Comment:
```suggestion
                # We may have one more epoch than necessary due to rounding:
```

Reply:
The original comment was incorrect but your proposal is not fully correct either:

```python
# %%
import numpy as np


num_of_epochs = 320
epoch_dur = 0.7
psd_signal_dur = 30

num_of_epochs_required = int(np.ceil(psd_signal_dur / epoch_dur))

epochs_idx = np.arange(
    start=0,
    stop=num_of_epochs,
    step=num_of_epochs // num_of_epochs_required
)

print(f'Required: {num_of_epochs_required}, got: {len(epochs_idx)}')
```

```
Required: 43, got: 46
```

This is because we use a floored `step` width.

If you have any idea how to make this better, let me know!

Reply:
how about this:

`epochs_idx = np.unique(np.round(np.linspace(0, num_of_epochs-1, num_of_epochs_required)).astype(int))`

?

Reply:
Due to the `unique()` we might end up with fewer epochs than desired in some cases, no?

Reply:
yes but to be honest I would do this. I would suggest that psd_signal_duration means up to psd_signal_duration seconds. So no need to raise an exception and it won't crash randomly on some subjects with many dropped epochs...

Reply:
I just played with some toy data and found that `np.unique()` dropping one of our epochs is really an absolute edge case, so I'm now using your code and emit a warning if the resulting duration is less than what the user had requested. I think this should be good for the vast majority of users.